test/bin.spec.js

```diff
@@ -13,6 +34,94 @@ test('w3 --version', (t) => {
 })
 
 test('w3 space create', (t) => {
-  const { stdout } = execaSync('./bin.js', ['space', 'create'])
+  const { stdout } = execaSync('./bin.js', ['space', 'create'], { env: createEnv() })
   t.regex(stdout, /^did:key:/)
 })
+
+test('w3 up', async (t) => {
+  await execa('./bin.js', ['space', 'create'], { env: createEnv() })
+
+  const service = mockService({
+    store: {
+      add: provide(StoreCapabilities.add, () => ({
+        status: 'upload',
+        headers: { 'x-test': 'true' },
+        url: 'http://localhost:9200'
+      }))
+    },
+    upload: {
+      add: provide(UploadCapabilities.add, ({ invocation }) => {
+        const { nb } = invocation.capabilities[0]
+        if (!nb) throw new Error('missing nb')
+        return nb
+      })
+    }
+  })
+
+  const serviceSigner = await Signer.generate()
+  const server = createServer({
+    id: serviceSigner,
+    service,
+    decoder: CAR,
+    encoder: CBOR
+  })
+
+  t.context.setRequestListener(createHTTPListener(server))
+
+  const env = createEnv(serviceSigner, t.context.serverURL)
+  const { stderr } = await execa('./bin.js', ['up', 'test/fixtures/pinpie.jpg'], { env })
+
+  t.true(service.store.add.called)
+  t.is(service.store.add.callCount, 1)
+  t.true(service.upload.add.called)
+  t.is(service.upload.add.callCount, 1)
+
+  t.regex(stderr, /Stored 1 file/)
+})
+
+test('w3 ls', async (t) => {
+  await execa('./bin.js', ['space', 'create'], { env: createEnv() })
+
+  const uploads = []
+
+  const service = mockService({
+    store: {
+      add: provide(StoreCapabilities.add, () => ({
+        status: 'upload',
+        headers: { 'x-test': 'true' },
+        url: 'http://localhost:9200'
+      }))
+    },
+    upload: {
+      add: provide(UploadCapabilities.add, ({ invocation }) => {
+        const { nb } = invocation.capabilities[0]
+        if (!nb) throw new Error('missing nb')
+        uploads.push(nb)
+        return nb
+      }),
+      list: provide(UploadCapabilities.list, () => {
+        return { results: uploads, size: uploads.length }
+      })
+    }
+  })
+
+  const serviceSigner = await Signer.generate()
+  const server = createServer({
+    id: serviceSigner,
+    service,
+    decoder: CAR,
+    encoder: CBOR
+  })
+
+  t.context.setRequestListener(createHTTPListener(server))
+
+  const env = createEnv(serviceSigner, t.context.serverURL)
+
+  const list0 = await execa('./bin.js', ['ls'], { env })
+  t.regex(list0.stdout, /No uploads in space/)
+
+  await execa('./bin.js', ['up', 'test/fixtures/pinpie.jpg'], { env })
+
+  const list1 = await execa('./bin.js', ['ls'], { env })
+  t.notThrows(() => CID.parse(list1.stdout.trim()))
```

Comment:
Might be better to pass --json here. We might change human friendly output of the ls command, but the (nd)json form is less likely to change.

Reply:
...or at least we're more likely to keep it backwards compatible